src/QsCompiler/Core/Transformations/ExpressionTransformation.fs

```diff
@@ -17,51 +17,36 @@ open Microsoft.Quantum.QsCompiler.Transformations.Core.Utils
 
 type private ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>
 
+type ExpressionKindTransformationBase(expressionTransformation: _ -> ExpressionTransformationBase, options) =
+    static let createExpressionTransformation kinds options =
+        let types = TypeTransformationBase options
+        let expressions = ExpressionTransformationBase((fun () -> kinds), (fun () -> types), options)
+        fun () -> expressions
 
-type ExpressionKindTransformationBase internal (options: TransformationOptions, _internal_) =
+    let node = if options.Rebuild then Fold else Walk
 
-    let missingTransformation name _ =
-        new InvalidOperationException(sprintf "No %s transformation has been specified." name) |> raise
+    member _.Types = expressionTransformation().Types
 
-    let Node = if options.Rebuild then Fold else Walk
+    member _.Expressions = expressionTransformation ()
 
-    member val internal ExpressionTransformationHandle = missingTransformation "expression" with get, set
-
-    member this.Expressions: ExpressionTransformationBase = this.ExpressionTransformationHandle()
-    member this.Types: TypeTransformationBase = this.ExpressionTransformationHandle().Types
-    member this.Common: CommonTransformationNodes = this.ExpressionTransformationHandle().Types.Common
-
-    new(expressionTransformation: unit -> ExpressionTransformationBase, options: TransformationOptions) as this =
-        new ExpressionKindTransformationBase(options, "_internal_")
-        then this.ExpressionTransformationHandle <- expressionTransformation
+    member internal this.Common = expressionTransformation().Types.Common
 
     // TODO: RELEASE 2022-09: Remove member.
     [<Obsolete("Please use ExpressionKindTransformationBase(unit -> ExpressionTransformationBase, TransformationOptions) instead.")>]
-    new(expressionTransformation: unit -> ExpressionTransformationBase,
-        typeTransformation: unit -> TypeTransformationBase,
-        options: TransformationOptions) = new ExpressionKindTransformationBase(expressionTransformation, options)
-
-    new(options: TransformationOptions) as this =
-        new ExpressionKindTransformationBase(options, "_internal_")
-        then
-            let typeTransformation = new TypeTransformationBase(options)
-
-            let expressionTransformation =
-                new ExpressionTransformationBase((fun _ -> this), (fun _ -> typeTransformation), options)
+    new(expressionTransformation, typeTransformation: unit -> TypeTransformationBase, options: TransformationOptions) =
+        ExpressionKindTransformationBase(expressionTransformation, options)
```

Comment:
We should be using signature files to give explicit types to the public API of the transformations, but I don't want to add inline type annotations to the .fs file unless necessary to resolve ambiguous types, because IMO it adds clutter and is harder to read.

Reply:
I really don't agree with that direction for the style, but since this is subjective, I recommend we leave the type annotations that are already present in the code, while we discuss as a broader team discussion about this style.

Reply:
When discussing F# with other developers, they have consistently told me that they use type annotations for function parameters, as opposed to using signature files. I am not in favor of using signature files per my experience with them in the Q# formatter project. They made working in that code base significantly harder and created an even further divorce of the parameters from the types that describe them. Since we are in disagreement on this, I appreciate that I can't ask you to make a commitment to not having type annotations in general, but as a compromise I am asking you to leave the type annotations that already exist in the code as they are.

Reply:
We've already had internal team discussions about the use of F# signature files, and decided in favor of using them. They are recommended by the [F# component design guidelines](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/component-design-guidelines#consider-using-explicit-signature-files-fsi-for-stable-library-and-component-apis) which our F# code should follow. Conversely, the [F# coding conventions](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/conventions) do not recommend that type annotations should be used for parameter types, other than to specify the public API which is accomplished by signature files.

Please follow up with me in private if you need more clarity on our team's F# coding guidelines.